### PR TITLE
Default XLA to use swap_tensors path in nn.Module._apply

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126984
* #126788
* #126819
* __->__ #126814
* #127313

### CODE REVIEW

**Review by JackCaoG — APPROVED** [review]

**Comment by JackCaoG** [comment]
Thanks @mikaylagawarecki

File: test/test_nn.py [hunk 1]
```diff
@@ -8174,9 +8174,9 @@ def test_batchnorm_large_batch(self, device, dtype):
     @dtypes(torch.float, torch.double, torch.bfloat16, torch.complex128)
     def test_conv_empty_input(self, device, dtype):
         def help(input, conv, memory_format):
-            ref_out = conv(input)
+            ref_out = conv(input).detach()
```
Comment (mikaylagawarecki) [inline test/test_nn.py:8177]: These `.detach()` calls are to ensure the autograd graph is not alive during `.to()` which will increase the `refcount` of params and prevent `swap_tensors` from being used.

As discussed, this was a known limitation of the `swap_tensors` path, but is unlikely to occur in practice (you don't normally want to change the dtype/device of your model after running forward).

Comment (JackCaoG) [reply test/test_nn.py:8177]: yea I agree that in real life it is unlikely to happen

**Review by albanD — APPROVED** [review]

**Comment by JackCaoG** [comment]
Sorry my bad, upstream runs a subset of the full XLA test. I started to see the CI failure on our end with
```
======================================================================
ERROR: test_sync_bn1d_multi_channel (__main__.TestMpSyncBatchNorm)
TestMpSyncBatchNorm.test_sync_bn1d_multi_channel
----------------------------------------------------------------------
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 826, in _apply
    torch.utils.swap_tensors(param, param_applied)
  File "/usr/local/lib/python3.10/site-packages/torch/utils/__init__.py", line 68, in swap_tensors
    torch._C._swap_tensor_impl(t1, t2)
RuntimeError: Expected single reference to a's Tensor object but got 2

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/usr/local/lib/python3.10/concurrent/futures/process.py", line 205, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/usr/local/lib/python3.10/concurrent/futures/process.py", line 205, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/usr/local/lib/python3.10/site-packages/torch_xla/runtime.py", line 95, in wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/torch_xla/_internal/pjrt.py", line 78, in _run_thread_per_device
    replica_results = list(
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 621, in result_iterator
    yield _result_or_cancel(fs.pop())
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 319, in _result_or_cancel
    return fut.result(timeout)
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/site-packages/torch_xla/_internal/pjrt.py", line 71, in _thread_fn
    return fn()
  File "/usr/local/lib/python3.10/site-packages/torch_xla/_internal/pjrt.py", line 190, in __call__
    self.fn(runtime.global_ordinal(), *self.args, **self.kwargs)
  File "/__w/xla/xla/pytorch/xla/test/test_mp_sync_batch_norm.py", line 87, in _sync_bn1d_multi_channel
    assert_stats(sbn_xla.cpu(), bn_cpu)
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 972, in cpu
    return self._apply(lambda t: t.cpu())
  File "/usr/local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 830, in _apply
    raise RuntimeError(f"_apply(): Couldn't swap ***self._get_name()***.***key***") from e
RuntimeError: _apply(): Couldn't swap SyncBatchNorm.weight
"""
```
It is ok to revert this or while I debug this issue? Thanks!

**Comment by mikaylagawarecki** [comment]
@JackCaoG curious, is that the only test that fails or are there more tests, I expect this to happen for all tests where the autograd graph is still alive when calling `._apply`, in the above case, I think this line is the line keeping the autograd graph alive https://github.com/pytorch/xla/blob/7350b7029a75b248bdbacadde20baaf3ccbbb4ef/test/test_mp_sync_batch_norm.py#L79

**Comment by JackCaoG** [comment]
I think all tests in `xla/test/test_mp_sync_batch_norm.py` failed. So far this seems like the only test failed. 

@mikaylagawarecki Would you mind help me understand the `autograd graph is still alive` part? From the test code it seems like `backward` is already being called in https://github.com/pytorch/xla/blob/7350b7029a75b248bdbacadde20baaf3ccbbb4ef/test/test_mp_sync_batch_norm.py#L20-L22. I thought autograd graph should be gone after backward.

**Comment by JackCaoG** [comment]
Thanks! That make a lot of sense. I am slightly concern about this behavior through. It seems like we can't move the module back to the CPU after the fwd/bwd is run, even from the user perspective the autograd state is "clean". Is there a way around this issue?

Sometimes user will want to move the module to the CPU and then apply checkpoint.

**Comment by mikaylagawarecki** [comment]
@JackCaoG I just merged https://github.com/pytorch/pytorch/pull/12731 that addresses the issue in the comment above

If `use_count = 2`, the `swap_tensors` call will now check whether the second reference is held by the `AccumulateGrad` node, if so it will register a prehook to the `AccumulateGrad` node that marks it as unusable. After the device/dtype changes  there will be a new AccumulateGrad node, so we never expect to touch this AccumulateGrad node again when running backwardd

Changing the device/dtype after running forward+backward should now always run without erroring. Does this resolve the errors you ran into in your CI?

```python
import torch
import torch.nn as nn
torch.__future__.set_swap_module_params_on_conversion(True)
m = nn.Linear(2, 3, device='cuda')
inp = torch.randn(2, 2, device='cuda')
out = m(inp)
out.sum().backward()
m.cpu()
inp2 = torch.randn(2, 2)
out2 = m(inp2)
out2.sum().backward()
print(m.weight.grad)
```

Also, I want to correct what I said about changing the device/dtype after running `forward` (but before backward), this is not  ok as in addition to the AccumulateGrad node, autograd might save a reference for backward we see this (e.g. for Conv but not for Linear)

**Comment by JackCaoG** [comment]
Ok there were some issue with XLA's CI when it is pin to a upstream pr. I manually test your pr and seems like `test_mp_sync_batch_norm.py` is passing now. I think you can merge this one. Thanks a lot!

**Comment by mikaylagawarecki** [comment]
@izaitsevfb hm weird indeed the regression seems to be XNNpack related but this diff does not touch XNNpack, reimported as D58094197 to re-run the test

**Comment by mikaylagawarecki** [comment]
@izaitsevfb I no longer see the failure in [D58015016](https://www.internalfb.com/diff/D58015016) in the new import D58094197, is this okay to merge?

**Comment by mikaylagawarecki** [comment]
[linux-binary-manywheel / manywheel-py3_8-cuda11_8-test / test](https://hud.pytorch.org/pr/pytorch/pytorch/126814#25804280149) ([gh](https://github.com/pytorch/pytorch/actions/runs/9371122018/job/25804280149))
ImportError: libcudnn.so.9: cannot open shared object file: No such file or directory
[trunk / macos-13-py3-arm64 / build](https://hud.pytorch.org/pr/pytorch/pytorch/126814#25805954671) ([gh](https://github.com/pytorch/pytorch/actions/runs/9371122025/job/25805954671))
/Users/ec2-user/runner/_work/pytorch/pytorch/c10/util/StringUtil.cpp:45:8: error: 'wstring_convert<std::codecvt_utf8_utf16<wchar_t>>' is deprecated [-Werror,-Wdeprecated-declarations]

failures are unrelated

**Pull request closed without merging**